module/extension/Director/src/Director.cpp

```diff
@@ -0,0 +1,253 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2021 NUbots <nubots@nubots.net>
+ */
+#include "Director.hpp"
+
+#include "extension/Configuration.hpp"
+
+namespace module::extension {
+
+    using ::extension::Configuration;
+    using ::extension::behaviour::Task;
+    using ::extension::behaviour::commands::CausingExpression;
+    using ::extension::behaviour::commands::DirectorTask;
+    using ::extension::behaviour::commands::ProviderClassification;
+    using ::extension::behaviour::commands::ProviderDone;
+    using ::extension::behaviour::commands::ProvidesReaction;
+    using ::extension::behaviour::commands::WhenExpression;
+    using ::NUClear::threading::Reaction;
+    using Unbind = NUClear::dsl::operation::Unbind<ProvidesReaction>;
+
+    using TaskPack = std::vector<std::shared_ptr<const DirectorTask>>;
+
+    /**
+     * This message gets emitted when a state we are monitoring is updated.
+     * When a `When` condition is being waited on a reaction will start monitoring that state.
+     * When it updates we check to see if this would now satisfy the conditions and if it does we emit
+     */
+    struct StateUpdate {
+        StateUpdate(const uint64_t& provider_id_, const std::type_index& type_, const int& state_)
+            : provider_id(provider_id_), type(type_), state(state_) {}
+
+        /// The reaction id of the Provider which was waiting on this type
+        uint64_t provider_id;
+        /// The enum type that this enum was listening for
+        std::type_index type;
+        /// The new value for the enum
+        int state;
+    };
+
+    /**
+     * Adds a Provider to the list of active Providers
+     *
+     * @param provides the provides message that contains the details we need to make a new provider
+     *
+     * @throws std::runtime_error if `r` has previously had a Provider added
+     */
+    void Director::add_provider(const ProvidesReaction& provides) {
+
+        auto& group = groups[provides.type];
+
+        // Check if we already inserted this element as a Provider
+        if (providers.count(provides.reaction->id) != 0) {
+            throw std::runtime_error("You cannot have multiple Provider DSL words in a single on statement.");
+        }
+
+        auto provider = std::make_shared<Provider>(provides.classification, provides.type, provides.reaction);
+        group.providers.push_back(provider);
```

Comment:
I think it should be mentioned that `push_back` will only accept objects of the expected type, or objects which can be casted to said type, and the new element in the vector with either be copy or move constructed (via the appropriate constructor). `emplace_back` will pass its arguments to a constructor of the expected type and will therefore "construct in place". Having said this, the copy and move constructors are also constructors that `emplace_back` can call and so I don't see any reason why we should use `push_back` in this instance?

Reply:
@KipHamiltons

Reply:
Because we're not using `emplace_back` as the constructor. My understanding (stolen shamelessly from man with big brain  [here](https://www.youtube.com/watch?v=jKS9dSHkAZY) is that `push_back` is good for when you have already constructed the object, whereas `emplace_back` is when you're constructing in place. There's certainly a case to be made here for removing the binding to `provider` and constructing it in the `emplace_back`, but as it stands, if you're going to construct it first, I think it should be `push`

Reply:
Usually the compiler will optimise and omit the copy anyway.